server/hwapi/data_models/repository.py

```diff
@@ -107,3 +111,253 @@ def get_or_create(
+            models.Device.category.in_(
+                [DeviceCategory.NETWORK.value, DeviceCategory.OTHER]
+            ),
+        ),
+    ).first()
+
+
+def find_matching_wireless_device(
+    db: Session,
+    devices_query: Query,
+    wireless_validator: device_validators.WirelessAdapterValidator,
+) -> models.Device | None:
+    """
+    Find a network wireless device across the given list of devices that has the same
+    identifier and model name
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == wireless_validator.identifier.lower(),
+            models.Device.name == wireless_validator.model,
+            models.Device.category.in_([DeviceCategory.WIRELESS, DeviceCategory.OTHER]),
+        ),
+    ).first()
+
+
+def find_matching_audio_device(
+    db: Session,
+    devices_query: Query,
+    audio_validator: device_validators.AudioValidator,
+) -> models.Device | None:
+    """
+    Find a network audio device across the given list of devices that has the same
+    identifier and model name
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == audio_validator.identifier.lower(),
+            models.Device.name == audio_validator.model,
+            models.Device.category.in_([DeviceCategory.AUDIO, DeviceCategory.OTHER]),
+        ),
+    ).first()
+
+
+def find_matching_capture_device(
+    db: Session,
+    devices_query: Query,
+    capture_validator: device_validators.VideoCaptureValidator,
+) -> models.Device | None:
+    """
+    Find a network video capture device across the given list of devices that
+    has the same identifier and model name
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == capture_validator.identifier.lower(),
+            models.Device.name == capture_validator.model,
+            models.Device.category.in_([DeviceCategory.CAPTURE, DeviceCategory.OTHER]),
+        ),
+    ).first()
+
+
+def find_matching_pci_device(
+    db: Session,
+    devices_query: Query,
+    pci_validator: device_validators.PCIPeripheralValidator,
+) -> models.Device | None:
+    """
+    Find an arbitary pci device across the given list of devices that has the same
+    PCI ID and model name
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == pci_validator.pci_id.lower(),
+            models.Device.bus == BusType.pci,
+            models.Device.name == pci_validator.name,
+        ),
+    ).first()
+
+
+def find_matching_usb_device(
+    db: Session,
+    devices_query: Query,
+    usb_validator: device_validators.USBPeripheralValidator,
+) -> models.Device | None:
+    """
+    Find an arbitary usb device across the given list of devices that has the same
+    USB ID and model name
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == usb_validator.usb_id.lower(),
```

Comment:
This is a generic question. What is `identifier`?

Reply:
I had the same question! I thought I was being dense and I eventually convinced myself that it was likely the PCI-ID, USB-ID, etc. dependent on the device? I am curious if that's true now though, and agree it would be nice to make this more clear, either in the naming or in some documentation

Reply:
yeah, it's PCI or USB ID. We use the same naming convension in C3